### PR TITLE
[IMP] base_exception: Add a hook method / Don't build void exceptions

### PR DESCRIPTION
**Pull request by rousseldenis**
In order to get exceptions without writing recordset, add a hook method.

Improve exceptions dictionaries building using a defaultdict of a browse record.

### CODE REVIEW

File: base_exception/models/base_exception_method.py [hunk 1]
```diff
@@ -35,35 +35,42 @@ def _rule_domain(self):
...
             to_remove = records_with_rule_in_exceptions - records_with_exception
             to_add = records_with_exception - records_with_rule_in_exceptions
-            # we expect to always work on the same model type
-            if rule_info.id not in rules_to_remove:
-                rules_to_remove[rule_info.id] = main_records.browse()
-            rules_to_remove[rule_info.id] |= to_remove
-            if rule_info.id not in rules_to_add:
-                rules_to_add[rule_info.id] = main_records.browse()
-            rules_to_add[rule_info.id] |= to_add
+            if to_remove:
```
Comment (hparfr) [inline base_exception/models/base_exception_method.py:61]: does this `if` improve something ?

**Review by hparfr — APPROVED** [review]

**Comment by hparfr** [comment]
by the way, what's your use case for this hook ?

**Comment by rousseldenis** [comment]
> by the way, what's your use case for this hook ?

Because we changed on a model the field exception_ids to a computed stored one for performance reasons and we don't want to do writes() in a compute.

IMHO, some improvements should be done in order to use compute/stored fields instead doing detect_exceptions in write/create. The problem is the trigger.

File: base_exception/models/base_exception_method.py (hunk 1 above)
Comment (rousseldenis) [reply base_exception/models/base_exception_method.py:61]: Yes, the dict will be lighter as currently containing all the rules.

**Review by lmignon — APPROVED** [review]

**Pull request merged**